Tooling: Add CI-aligned devcontainer; See #7906

mgajek-cern opened the pull request:
Closes #7906 

Major changes:

 - Add .devcontainer/alma9/ with devcontainer.json using rucio-dev:release-38.0.0
    - Configure VS Code Python extension and Node.js/Git features
    - Add common_functions.sh with Docker/system checks, service management, and CI command display
    - Add initialize.sh to start external services with configurable profiles (storage, externalmetadata, etc.)
    - Add setup.sh to install dev tools (ruff, flake8, pytest, pyright) and Rucio in editable mode
    - Add help.sh showing CI-aligned commands matching GitHub Actions workflows
    - Support profile customization via rucio_dev_profiles.cfg
    - Mount Docker socket, enable privileged mode, and connect container to dev_default network
    - Set PYTHONPATH, RUCIO_HOME, and RUCIO_POLICY_PACKAGE environment variables
 
Improves developer onboarding and ensures local testing matches CI environment.

---

Screenshots of dev container setup in VS Code:

- Screenshots on `docker exec -t dev-rucio-1 tools/run_tests.sh -ir` command:
<img width="1465" height="907" alt="image" src="https://github.com/user-attachments/assets/102af63d-c785-43f6-aeeb-9a93f694f9be" />
<img width="1467" height="916" alt="image" src="https://github.com/user-attachments/assets/45dc98f0-8c5d-4cff-b961-f7387a010506" />

- Screenshot on `pytest tests/rucio --cov=lib/rucio`:
<img width="1468" height="916" alt="image" src="https://github.com/user-attachments/assets/96e1a002-ed0f-4bd9-979f-015cc11ed08c" />

- Screenshot on `Run Tests`:
<img width="1505" height="899" alt="image" src="https://github.com/user-attachments/assets/10feb323-df83-4991-85e5-22544049be38" />

- TODO(mgajek-cern): Video presenting evidence of debugging the server component and other components as outlined in https://github.com/rucio/rucio/pull/7953#issuecomment-3275297837

mgajek-cern commented:
I’ve added several potential reviewers here since I’m new and not sure who’s best for this topic. I think having 1 to 3 reviews should be enough to cover it. If anyone’s unable to review or prefers to be removed, just let me know and I’ll update the list. Thanks

rdimaio commented on .devcontainer/alma9/common_functions.sh:
```diff
@@ -0,0 +1,274 @@
+        echo -e "${GREEN} ✅ Sufficient disk space available${NC}"
+    fi
+    
+    # Check Docker daemon
+    echo -e "   Docker: Checking daemon status..."
+    if ! docker info > /dev/null 2>&1; then
+        echo -e "${RED} ❌ Docker daemon not accessible. Please ensure Docker is running.${NC}"
+        exit 1
+    else
+        echo -e "${GREEN} ✅ Docker daemon is running${NC}"
+    fi
+    echo ""
+}
+
+# Install Docker CLI for AlmaLinux
+install_docker_cli() {
+    echo -e "${BLUE}🐳 Installing Docker CLI and tools...${NC}"
+    if ! command -v docker &> /dev/null 2>&1; then
+        dnf config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
+        dnf install -y docker-ce-cli docker-compose-plugin
+        echo -e "${GREEN} ✅ Docker CLI installed${NC}"
+    else
+        echo -e "${GREEN} ✅ Docker CLI already available${NC}"
+    fi
+}
```
Could be worth using podman instead from the getgo here, since we sort of have support for it around Rucio, and in theory it'd be better if we relied on that instead of Docker

mgajek-cern replied:
Is the systemd/daemon support the main rationale for preferring Podman here, or are there other considerations (like existing Rucio infrastructure alignment, rootless containers, etc.) we should factor in?

The [documentation](https://github.com/rucio/rucio/blob/master/etc/docker/dev/README.rst) states:

> We provide a containerised version of the Rucio development environment for a quick start. Our containers are ready-made for Docker, but we provide best-effort support for Podman and podman-compose. If you use podman-compose, you need to add -t cntnet to the podman-compose commands. For simplicity all commands explained in this documentation are using docker-compose.

rdimaio replied:
I think people with more time in the project can chime in with a better reply than me here, but I think that at least in this context where it's a dev environment, podman might be preferable (even if just considering podman's open source policies compared to Docker's)

rdimaio commented on .devcontainer/alma9/common_functions.sh:
```diff
@@ -0,0 +1,274 @@
+    fi
+}
+
+cleanup_existing_containers() {
+    echo -e "${YELLOW}🔄 Cleaning up existing containers...${NC}"
+
+    # Remove containers prefixed with 'dev-'
+    echo -e "${YELLOW}🧹 Removing containers with prefix 'dev-'...${NC}"
+    containers=$(docker ps -a --filter "name=^dev-" --format "{{.ID}}")
+    if [ -n "$containers" ]; then
+        docker rm -f $containers
+        echo -e "${GREEN} ✅ Containers removed${NC}"
+    else
+        echo -e "${GREEN} ✅ No 'dev-' containers to remove${NC}"
+    fi
+
+    # Remove volumes prefixed with 'dev_'
+    echo -e "${YELLOW}🧹 Removing volumes with prefix 'dev_'...${NC}"
+    volumes=$(docker volume ls --filter "name=^dev_" --format "{{.Name}}")
+    if [ -n "$volumes" ]; then
+        docker volume rm $volumes
+        echo -e "${GREEN} ✅ Volumes removed${NC}"
+    else
+        echo -e "${GREEN} ✅ No 'dev_' volumes to remove${NC}"
+    fi
+
+    # Remove network named 'dev_default' if exists
+    echo -e "${YELLOW}🧹 Removing 'dev_default' network if exists...${NC}"
+    if docker network ls --filter "name=^dev_default$" --format "{{.Name}}" | grep -q 'dev_default'; then
+        docker network rm dev_default
+        echo -e "${GREEN} ✅ Network 'dev_default' removed${NC}"
+    else
+        echo -e "${GREEN} ✅ Network 'dev_default' not found${NC}"
+    fi
+
+    echo ""
+}
```
Assuming this is all within the devcontainer, is there no way to just clean up stuff within the devcontainer namespace?

mgajek-cern replied:
The devcontainer can't reliably handle docker-compose setup with complex volume mount paths for certificates and keys in `etc/docker/dev/docker-compose.yml` that must resolve correctly on the host filesystem (will reproduce and share screenshots if needed). We use `initializeCommand` to run the compose stack on the host before devcontainer startup, ensuring proper path resolution. 

Within the devcontainer itself, we access the host's Docker daemon through the mounted docker.sock, allowing container management from inside the devcontainer.

rdimaio commented on .devcontainer/alma9/devcontainer.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "Rucio alma9 Development Environment",
+  "image": "rucio/rucio-dev:release-38.0.0",
```
Maybe we should always point to latest here?

mgajek-cern replied:
We could configure Dependabot to automatically update the image version in devcontainer.json files. This would provide reproducible environments by default while ensuring automatic updates when new stable releases are available, rather than always pulling the potentially unstable latest tag.

This approach gives you both stability and currency - developers get consistent environments, but you don't fall behind on important updates.

rdimaio replied:
But in general I think this would need to be aligned with the Rucio version that's locally pulled, since I assume the dev would want to test their changes here

rdimaio reviewed (commented):
Overall this seems to useful and to make sense, but I'm only concerned about the effort that would be needed to maintain this logic in sync with the rest of the things - for example, maintaining the profiles here in sync with the Docker compose profiles, maintaining the integ test list in sync, etc. it would be great if these things could be put in some sort of single place where they can then be imported in both places, to avoid this issue.

I am also curious about the system requirements, particularly the disk space - 30+ GB for full testing seems excessive, did you test this?

I think the best person to review this would be @maany as he developed the Rucio VSCode dev environment https://github.com/maany/rucio-vscode-dev-env, so he might have a lot of context on these aspects

mgajek-cern commented:
> Overall this seems to useful and to make sense, but I'm only concerned about the effort that would be needed to maintain this logic in sync with the rest of the things - for example, maintaining the profiles here in sync with the Docker compose profiles, maintaining the integ test list in sync, etc. it would be great if these things could be put in some sort of single place where they can then be imported in both places, to avoid this issue.
> 
> I am also curious about the system requirements, particularly the disk space - 30+ GB for full testing seems excessive, did you test this?
> 
> I think the best person to review this would be @maany as he developed the Rucio VSCode dev environment https://github.com/maany/rucio-vscode-dev-env, so he might have a lot of context on these aspects

Thanks for the feedback @rdimaio. I'll try to sync with @maany this or subsequent week on the setup if feasible and share ideas on whether we should continue with these efforts

Geogouz commented:
My worries here is that I see too much overlapping functionality with the currently existing `bootstrap_dev.sh` which is on hold till our ongoing CI adjustments are ready, in order to be also adjusted if needed (see #7929). Currently that bootstrapping dev environment does work and is quite feature-full (but unfortunately its documentation lives at the moment only within the script itself). Nevertheless, this PR does offer couple features that are unique or follow nice practices.

Another point is that for a new developer, seeing a dev container under Rucio/containers, and yet, another dev container here, might bring questions.

So for the shake of keeping things minimal and concentrated, and avoiding to maintain two parallel bootstrapping paths that almost do the same thing (not to mention that Rucio's web-documentation would end up being quite awkward for the developers and unclear when to use what), I would suggest to keep a single approach by merging the best of those two worlds.

mgajek-cern commented:
> My worries here is that I see too much overlapping functionality with the currently existing `bootstrap_dev.sh` which is on hold till our ongoing CI adjustments are ready, in order to be also adjusted if needed (see #7929). Currently that bootstrapping dev environment does work and is quite feature-full (but unfortunately its documentation lives at the moment only within the script itself). Nevertheless, this PR does offer couple features that are unique or follow nice practices.
> 
> Another point is that for a new developer, seeing a dev container under Rucio/containers, and yet, another dev container here, might bring questions.
> 
> So for the shake of keeping things minimal and concentrated, and avoiding to maintain two parallel bootstrapping paths that almost do the same thing (not to mention that Rucio's web-documentation would end up being quite awkward for the developers and unclear when to use what), I would suggest to keep a single approach by merging the best of those two worlds.

I understand the overlap concern, but these serve fundamentally different workflows:

- `bootstrap_dev.sh`: Host-based container orchestration for testing/CI scenarios
- Devcontainer: Integrated IDE development environment with debugging, IntelliSense, and consistent workspace

Thanks for the feedback @Geogouz. The key differentiator is that devcontainers provide guaranteed debugging capabilities within the container environment, which `bootstrap_dev.sh` doesn't address. They're complementary rather than competing approaches.

To address the confusion concern, I could clearly document the distinct use cases and potentially reuse components from `bootstrap_dev.sh` within the devcontainer setup where appropriate.

Geogouz commented:
I would argue that both have basically the same host dependencies (namely, Docker daemon and Compose plugin). The current script spawns all the required containers on host, while the later spawns a single container on the host and then all others inside it. The practical difference I see (which is indeed a **very** nice feature to have out-of-the-box) is that with the proposed approach, VS Code (as I'm not sure if PyCharm fully supports `devcontainer.json` yet) attaches directly to the containerized workspace enabling breakpoints. By contrast, to use breakpoints now with the `bootstrap_dev.sh`, you would need to use something like `debugpy`.

So my dilemma boils down to.. assuming that we do expand `bootstrap_dev.sh` with a minimal commit, to automatically attach a debugger, does this PR still introduce new functionalities? And what about their cost/gain?
Being a PyCharm guy myself, I think ultimately @maany's opinion is more valuable than mine here (he is the VS Coder guru :D).

mgajek-cern commented:
> I would argue that both have basically the same host dependencies (namely, Docker daemon and Compose plugin). The current script spawns all the required containers on host, while the later spawns a single container on the host and then all others inside it. The practical difference I see (which is indeed a **very** nice feature to have out-of-the-box) is that with the proposed approach, VS Code (as I'm not sure if PyCharm fully supports `devcontainer.json` yet) attaches directly to the containerized workspace enabling breakpoints. By contrast, to use breakpoints now with the `bootstrap_dev.sh`, you would need to use something like `debugpy`.
> 
> So my dilemma boils down to.. assuming that we do expand `bootstrap_dev.sh` with a minimal commit, to automatically attach a debugger, does this PR still introduce new functionalities? And what about their cost/gain? Being a PyCharm guy myself, I think ultimately @maany's opinion is more valuable than mine here (he is the VS Coder guru :D).

You make a fair point about the architectural similarity and debugging capabilities. You're right that if we enhanced `bootstrap_dev.sh` with debugger attachment, the functional gap would narrow significantly.

The main remaining value would be the standardized devcontainer specification, which provides:
- Cross-platform IDE integration (VS Code native, PyCharm improving)
- Reproducible development environments without local setup variations
- Extension/tooling auto-installation

However, given the overlap you've identified and the ongoing CI work, I think the pragmatic approach is to keep this as a draft until `bootstrap_dev.sh` is finalized. Then we can make an informed decision about whether the devcontainer benefits justify maintaining both approaches or if we should integrate the best parts into a single solution.

I agree @maany's input as the VS Code expert will be crucial for this decision.

rdimaio commented:
>  (as I'm not sure if PyCharm fully supports `devcontainer.json` yet)


Seems like it does, even if just in PyCharm Pro https://www.jetbrains.com/help/pycharm/connect-to-devcontainer.html

Geogouz commented:
> The main remaining value would be the standardized devcontainer specification, which provides:
> 
> * Cross-platform IDE integration (VS Code native, PyCharm improving)
> * Reproducible development environments without local setup variations
> * Extension/tooling auto-installation

Fair points.

> Seems like it does, even if just in PyCharm Pro https://www.jetbrains.com/help/pycharm/connect-to-devcontainer.html

Gonna give it a try to see how things go with Pycharm too.

mgajek-cern commented:
After discussing with @maany, we've decided to:

- Wait for GitHub issue #7965 to be resolved before proceeding
- Integrate those changes once they're merged
- Expand the acceptance criteria and confirm that debugging works properly for tests, daemons, CLIs and server component (include video evidence)

Commit:
feat(devcontainer): Add CI-aligned AlmaLinux 9 development environment

- Add .devcontainer/alma9/ with devcontainer.json using rucio-dev:release-38.0.0
- Configure VS Code Python extension and Node.js/Git features
- Add common_functions.sh with Docker/system checks, service management, and CI command display
- Add initialize.sh to start external services with configurable profiles (storage, externalmetadata, etc.)
- Add setup.sh to install dev tools (ruff, flake8, pytest, pyright) and Rucio in editable mode
- Add help.sh showing CI-aligned commands matching GitHub Actions workflows
- Support profile customization via rucio_dev_profiles.cfg
- Mount Docker socket, enable privileged mode, and connect container to dev_default network
- Set PYTHONPATH, RUCIO_HOME, and RUCIO_POLICY_PACKAGE environment variables

Improves developer onboarding and ensures local testing matches CI environment.

mgajek-cern commented:
@Geogouz Depends on https://github.com/rucio/rucio/issues/7965. Work on this will continue once a PR to that issue is merged or the issue is closed with an explanation. I’ll rebase afterwards and proceed or someone else can take over if needed.

mgajek-cern commented:
Lets close this pull request based on yesterdays agreement related to DRAFT PRs which might depend on previous issues or PRs for now. We might re-open this in the future or address the issue in a separate PR
